lib/fs/spifs/spifs.c

```diff
@@ -0,0 +1,1146 @@
+    toc_file_t empty;
+    memset(&empty, 0, SPIFS_ENTRY_LENGTH);
+
+    spifs_file_t *file = list_peek_head_type(&spifs->files, spifs_file_t, node);
+    for (uint32_t i = 0; i < spifs->num_entries; i++) {
+        uint8_t *page_end = spifs->page + spifs->page_size;
+        DEBUG_ASSERT(cursor <= page_end);
+
+        if (cursor == page_end) {
+            err = spifs_write_page(spifs, toc_page_addr);
+            if (err != NO_ERROR) {
+                return err;
+            }
+
+            toc_page_addr += target_toc;
+            cursor = spifs->page;
+        }
+
+        if (file) {
+            crc = crc32(crc, (uint8_t*)&file->metadata, SPIFS_ENTRY_LENGTH);
+            memcpy(cursor, (uint8_t*)&file->metadata, SPIFS_ENTRY_LENGTH);
+            file = list_next_type(&spifs->files, &file->node, spifs_file_t, node);
+        } else {
+            crc = crc32(crc, (uint8_t*)&empty, SPIFS_ENTRY_LENGTH);
+            memcpy(cursor, (uint8_t*)&empty, SPIFS_ENTRY_LENGTH);
+        }
+
+        cursor += SPIFS_ENTRY_LENGTH;
+    }
+
+    // Sanity check. The cursor should be at the last position in this page
+    // at this point.
+    uint8_t* expected_cursor_location =
+        (spifs->page + spifs->page_size) - SPIFS_ENTRY_LENGTH;
+    DEBUG_ASSERT(cursor == expected_cursor_location);
+
+    toc_footer_t footer;
+    memset(&footer, 0, SPIFS_ENTRY_LENGTH);
+    footer.checksum = crc;
+    memcpy(cursor, (uint8_t*)&footer, SPIFS_ENTRY_LENGTH);
+
+    err = spifs_write_page(spifs, toc_page_addr);
+    if (err != NO_ERROR)
+        return err;
+
+    // Only update this once we're sure that the write went through.
+    // This way, if the write failed, we'll try writing over the bad ToC again
+    // rather than potentially corrupting both ToCs.
+    spifs->generation = target_generation;
+    spifs->toc_position = target_toc;
+
+    return NO_ERROR;
+}
+
+static void spifs_add_ascending(spifs_t *spifs, spifs_file_t *target)
+{
+    spifs_file_t *file;
+    list_for_every_entry(&spifs->files, file, spifs_file_t, node) {
+        if (file->metadata.page_idx > target->metadata.page_idx) {
+            list_add_before(&file->node, &target->node);
+            return;
+        }
+    }
+
+    list_add_tail(&spifs->files, &target->node);
+}
+
+
+static status_t spifs_read_page(spifs_t *spifs, uint32_t page_addr)
+{
+    off_t block_addr = page_addr * spifs->blocks_per_page;
+
+    ssize_t bytes = bio_read_block(spifs->dev, spifs->page, block_addr,
+                                   spifs->blocks_per_page);
+
+    if ((uint32_t)bytes != spifs->page_size) {
+        return ERR_IO;
+    }
+
+    return NO_ERROR;
+}
+
+static status_t spifs_write_page(spifs_t *spifs, uint32_t page_addr)
+{
+    off_t block_addr = page_addr * spifs->blocks_per_page;
+    off_t device_addr = block_addr * spifs->dev->block_size;
+
+    // Device requires erase before write?
+    if (spifs->dev->geometry_count != 0) {
```

Comment:
There was definitely some discussion but there doesn't appear to be any such flag in `bdev_t`